# Learning and practicing NumPy with examples and exercises

# Mi Cuaderno Interactivo: Explorando NumPy y Programación en Python

¡Bienvenidos! En este cuaderno, he reunido los puntos clave del taller que realicé, donde abordo:

- **Operaciones fundamentales con NumPy**: Para manipular y examinar datos de forma eficiente.

- **Ejemplo de descenso de gradiente en regresión lineal**: Para entender cómo se optimizan parámetros en problemas de aprendizaje automático.

Mi objetivo es mostrar, paso a paso, cómo aprovechar al máximo NumPy y sentar bases sólidas en programación y análisis de datos.

# **1 Ejemplos: NumPy**

## 1. Indexación Básica

Este ejemplo muestra cómo acceder a un elemento específico de un array unidimensional utilizando su índice.

In [9]:
import numpy as np

arr= np.array([10,20,30,40,50])
print(arr[2])


30


`arr[2]` accede al tercer elemento del array (**índice 2**), que es `30`.
## 2.Slicing
Este ejemplo demuestra cómo extraer una subsección de un array unidimensional utilizando slicing.

In [12]:
arrr=np.array([10,20,30,40,50])
print(arr[1:4])

[20 30 40]


`arr[1:4]` selecciona los elementos desde el **índice 1** hasta el **índice 3** (el **índice 4** no se incluye).

## 3.Indexación en Arrays 2D
Este ejemplo muestra cómo acceder a un elemento específico en un array bidimensional utilizando índices de fila y columna

In [14]:
arr= np.array([[1,2,3],[4,5,6],[7,8,9]])
print(arr[1,2])

6


`arr[1, 2]` accede al elemento en la segunda fila (**índice 1**) y tercera columna (**índice 2**), que es `6`.
## 4.Slicing en Arrays 2D
Este ejemplo muestra cómo extraer una submatriz de un array bidimensional utilizando slicing.

In [16]:
arr=np.array([[1,2,3],[4,5,6,],[7,8,9,]])
print(arr[:2,1:])

[[2 3]
 [5 6]]


`arr[:2, 1:]` selecciona una submatriz utilizando slicing:
- `:2` (para las filas): Selecciona las filas desde el inicio (**índice 0**) hasta la fila con **índice 1** (el **índice 2** no se incluye). Esto selecciona las dos primeras filas.
- `1:` (para las columnas): Selecciona las columnas desde el **índice 1** hasta el final. Esto selecciona las columnas 2 y 3.
 #### **resultado:** 

 $$
\begin{bmatrix}
2 & 3  \\
5 & 6 
\end{bmatrix}
$$

## 5.Indexación en Tensores 3D
Este ejemplo muestra cómo acceder a elementos específicos y submatrices en un tensor tridimensional.

In [17]:
tensor=np.array([[[1,2,3],[4,5,6,]],[[7,8,9],[13,11,12]],[[13,14,15],[16,17,18]]])
print(tensor[1,0,2])

9


- `tensor[1, 0, 2]` accede al elemento en la segunda matriz (**índice 1**), primera fila (**índice 0**), y tercera columna (**índice 2**), que es `9`.



In [18]:
print(tensor[1])

[[ 7  8  9]
 [13 11 12]]


- `tensor[1]` selecciona toda la segunda matriz (**índice 1**), que es 
$$
\begin{bmatrix}
7 & 8 & 9 \\
13 & 11 & 12
\end{bmatrix}
$$
## 6.Slicing en Tensores 3D
Este ejemplo muestra cómo extraer submatrices de un tensor tridimensional utilizando slicing.


In [33]:
tensor=np.arange(27).reshape(3,3,3)
print(tensor[1])




[[ 9 10 11]
 [12 13 14]
 [15 16 17]]


- `np.arange(27)` genera una lista de 27 numeros
- `.reshape(3,3,3)` convierte esa lista en un tensor, lo que tenemos son 3 matrices, cada una con 3 filas y 3 columnas.
- `tensor[1]` selecciona toda la segunda matriz (**índice 1**).


In [21]:
print(tensor[:,1,:])

[[ 3  4  5]
 [12 13 14]
 [21 22 23]]


- `tensor[:, 1, :]` selecciona la segunda fila (**índice 1**) de todas las matrices, formando una submatriz 3x3, que es 
\begin{bmatrix}
3& 4 & 5 \\
12 & 13 & 14\\
21 & 22 & 23
\end{bmatrix}
$$

In [22]:
print(tensor[:,:,2])

[[ 2  5  8]
 [11 14 17]
 [20 23 26]]


- `tensor[:, :, 1]` selecciona la tercera columna (**índice 2**) de todas las matrices, formando una submatriz 3x3, que es 
$$
\begin{bmatrix}
2 & 5 & 8 \\
11 & 14 & 17\\
20 & 23 & 26
\end{bmatrix}
$$

## 7. 4D Tensor Indexing
En este ejemplo se crea un tensor de 4 dimensiones y se accede a elementos y bloques específicos dentro de él.

In [35]:
tensor=np.arange(48).reshape(2,3,2,4)
print(tensor[1,2,0,3])


43


- `np.arange(48)` genera una lista de numeros del 0 al 47
- `reshape(2,3,2,4)` le da forma de tensor de 4 dimensiones: 2 bloques, 3 filas, 2 columnas y 4 valores por fila
- `tensor[1, 2, 0, 3]` accede al valor ubicado en el segundo bloque, tercera fila, primera subfila, cuarta posición → da como resultado `43`.

In [36]:
print(tensor[1,2])

[[40 41 42 43]
 [44 45 46 47]]


- `tensor[1, 2]` accede a todo el bloque en la posición [1, 2], devolviendo una matriz de **2x4**.

\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
$$

## 8.Slicing a 4D Tensor
En este ejemplo se realiza slicing sobre un tensor 4D para obtener subconjuntos a lo largo de diferentes dimensiones.

In [ ]:
tensor=np.arange(48).reshape(2,3,2,4)
print(tensor[1])


[[[[ 0  1  2  3]
   [ 4  5  6  7]]

  [[ 8  9 10 11]
   [12 13 14 15]]

  [[16 17 18 19]
   [20 21 22 23]]]


 [[[24 25 26 27]
   [28 29 30 31]]

  [[32 33 34 35]
   [36 37 38 39]]

  [[40 41 42 43]
   [44 45 46 47]]]]
[[[24 25 26 27]
  [28 29 30 31]]

 [[32 33 34 35]
  [36 37 38 39]]

 [[40 41 42 43]
  [44 45 46 47]]]


- `tensor[1]` selecciona el segundo bloque completo (de forma 3x2x4).
$$
\left[
\begin{array}{c}
\begin{bmatrix}
24 & 25 & 26 & 27 \\
28 & 29 & 30 & 31
\end{bmatrix} \\
\\
\begin{bmatrix}
32 & 33 & 34 & 35 \\
36 & 37 & 38 & 39
\end{bmatrix} \\
\\
\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
\end{array}
\right]
$$





In [41]:
print(tensor[:,2,:,:])

[[[16 17 18 19]
  [20 21 22 23]]

 [[40 41 42 43]
  [44 45 46 47]]]


- `tensor[:, :, 0, :]` selecciona la primera subfila (dimensión 3) de cada fila de cada bloque → forma final **2x3x4**.
$$
\left[
\begin{array}{c}
\begin{bmatrix}
16 & 17 & 18 & 19 \\
20 & 21 & 22 & 23
\end{bmatrix} \\
\\
\begin{bmatrix}
40 & 41 & 42 & 43 \\
44 & 45 & 46 & 47
\end{bmatrix}
\end{array}
\right]
$$

## 9. Matrix Addition
En este ejemplo se suman dos matrices del mismo tamaño elemento por elemento.

In [42]:
A=np.array([[1,2],[3,4]])
B=np.array([[5,6],[7,8]])
print(A+B)

[[ 6  8]
 [10 12]]


- Se definen dos matrices 2x2: **A** y **B**.

- La suma A + B suma cada elemento correspondiente:

`[[1+5, 2+6], [3+7, 4+8]] → [[6, 8], [10, 12]]`

- El resultado es una matriz del mismo tamaño con los valores sumados.

## 10. Matrix Multiplication
En este ejemplo se realiza la multiplicación matricial (producto punto) entre dos matrices compatibles.

In [44]:
A=np.array([[1,2],[3,4]])
B=np.array([[5,6],[7,8]])
result=np.dot(A,B)
print(result)

[[19 22]
 [43 50]]


- `np.dot(A, B)` aplica el producto punto entre matrices:

La forma de ambas es 2x2, por lo que son compatibles.

Resultado:
\begin{bmatrix}
19 & 22  \\
43 & 50
\end{bmatrix}
$$

## 11. Transpose of a Matrix
En este ejemplo se obtiene la transpuesta de una matriz, intercambiando filas por columnas.

In [45]:
A=np.array([[1,2],[3,4]])
result=np.transpose(A)
print(result)

[[1 3]
 [2 4]]


- `np.transpose(A)` invierte las filas y columnas de la matriz **A**.

Original:
$$
\begin{bmatrix}
1 & 2 \\
3 & 4 
\end{bmatrix}
$$

Transpuesta:
$$
\begin{bmatrix}
1 & 3 \\
2 & 4 
\end{bmatrix}
$$


## 12. Matrix Determinant and Inverse
En este ejemplo se calcula el determinante y la inversa de una matriz cuadrada.

In [48]:
A=np.array([[1,2],[3,4]])
det=np.linalg.det(A)
inverse=np.linalg.inv(A)
print('Determinant:',det)
print('Inverse: ',inverse)

Determinant: -2.0000000000000004
Inverse:  [[-2.   1. ]
 [ 1.5 -0.5]]


- `np.linalg.det(A)` calcula el determinante de **A**:
`1×4 - 2×3 = -2`

- `np.linalg.inv(A)` calcula la matriz inversa

Resultado:
$$
\begin{bmatrix}
-2 & 1 \\
1.5 & -0.5 
\end{bmatrix}
$$

## 13. Eigenvalues and Eigenvectors
En este ejemplo se calculan los valores propios **(**eigenvalues**) y vectores propios (**eigenvectors**) de una matriz cuadrada.

In [49]:
A=np.array([[1,-1],[1,3]])
aigenvalues, eigenvectors = np.linalg.eig(A)
print('Eigenvalues:', aigenvalues)
print('Eigenvectors:', eigenvectors)

Eigenvalues: [2.00000002 1.99999998]
Eigenvectors: [[-0.70710677 -0.70710679]
 [ 0.70710679  0.70710677]]


- `np.linalg.eig(A)`devuelve:

- `eigenvalues`: una lista con los valores propios λ.

- `eigenvectors`: una matriz donde cada columna es el vector propio correspondiente a cada 
𝜆.

## 14. Solving Linear Equations
En este ejemplo se resuelve un sistema de ecuaciones lineales de la forma $$ Ax= b $$.

In [50]:
A=np.array([[2,1],[1,1]])
b=np.array([3,2])
solution=np.linalg.solve(A,b)
print('Solution:',solution)

Solution: [1. 1.]


- Se tiene un sistema de ecuaciones lineales:

$$
\begin{cases}
2x + y = 3 \\
x + y = 2
\end{cases}
$$

- `np.linalg.solve(A, b)` resuelve este sistema usando álgebra lineal.

- El resultado es el valor de \( x \) que satisface ambas ecuaciones:

$$
x = \begin{bmatrix} 1 \\ 1 \end{bmatrix}
$$


## 15. Gradient Descent Optimization
En este ejemplo se aplica el algoritmo de descenso del gradiente para ajustar parámetros y minimizar el error cuadrático medio (MSE) de un modelo de regresión lineal.

In [51]:
np.random.seed(0)
x=2*np.random.rand(100,1)
y=4+3*x+np.random.randn(100,1)

- Se generan 100 datos aleatorios para x entre 0 y 2.

- Se genera `y` como una línea recta: $$y=4+3x+ruido$$


- El ruido se agrega para simular variabilidad real.

In [52]:
x_b=np.c_[np.ones((100,1)),x]


- Se agrega una columna de unos a `x` para incluir el término independiente (intercepto) en el modelo.

- Resultado: una matriz de forma (100, 2) donde la primera columna es 1 y la segunda es `x`.

In [53]:
learning_rate=0.1
n_iterations=1000

- `learning_rate`: tasa de aprendizaje.

- `n_iterations`: número de iteraciones del ciclo de entrenamiento.

In [55]:
theta = np.random.randn(2, 1)

- Se inicializa aleatoriamente el vector de parámetros 
**theta** con forma (2, 1).

In [57]:
for iteration in range(n_iterations):
 gradients = 2 / 100 * x_b.T.dot(x_b.dot(theta)- y)
 theta = theta- learning_rate * gradients
print("Final theta:", theta)

Final theta: [[4.22215108]
 [2.96846751]]



- Calcula el gradiente del error cuadrático medio.

- Actualiza los parámetros theta moviéndose en la dirección opuesta al gradiente.

- Se repite 1000 veces.
- El modelo aprendió una línea recta cercana a:
$$y=4.22+2.97x$$